### PR TITLE
Add server-friendly content sanitizer for Plugins SSR

### PR DESCRIPTION
**Pull request by claudiucelfilip**
#### Proposed Changes

The current content sanitizer for Plugins doesn't work on the server. This implements a similar API that works on the server using a 3rd party node module:`jsdom`.

#### Testing Instructions

* Log out and disable JavaScript
* Go to /plugins
* You should still be able to see the Plugins landing page with the 3 sections Paid, Editor's Pick, and Free
* The unit tests from `client/lib/plugins/sanitize-section-content.js` should pass

* Enable JavaScript
* Visit a few Plugin Details page to check the content is still shown

### CODE REVIEW

File: client/lib/plugins/server-dom-parser.js
Comment (tyxla) [inline client/lib/plugins/server-dom-parser.js]: Let's make sure to explain what `4294967295` is.

File: client/lib/plugins/server-dom-parser.js
Comment (tyxla) [inline client/lib/plugins/server-dom-parser.js]: Any chance one of those parsing functions could error and throw? Should we `try/catch` such cases just in case?

File: client/lib/plugins/test/server-dom-parser.js
Comment (tyxla) [inline client/lib/plugins/test/server-dom-parser.js]: This test made me wonder: where are the disallowed tags and attributes defined? Any way we can get it work as close as possible to what `wp_kses()` does?

**Review by tyxla — COMMENTED** [review]
Thanks for working on this 👍 

I wonder if we should ping some secops folks for a review, just in case.

I've also left a few questions, let me know what you think.

File: client/lib/plugins/server-dom-parser.js
Comment (claudiucelfilip) [reply client/lib/plugins/server-dom-parser.js]: Sounds good. Thanks!
It'll now return null and skip sanitization if something throws.

File: client/lib/plugins/test/server-dom-parser.js
Comment (claudiucelfilip) [reply client/lib/plugins/test/server-dom-parser.js]: Good point! they were defined through a couple of switches ([here](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/plugins/sanitize-section-content.js#L16) and [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/plugins/sanitize-section-content.js#L57)).
I replaced them with something similar to the `wp_kses()` allowed tags configurations.

**Comment by claudiucelfilip** [comment]
> I wonder if we should ping some secops folks for a review, just in case.

@tyxla  I reached out to them (in p1664528549877929-slack-C4EUDHL80) and @xknown provided a couple sanitization bypasses for the server-dom. The original client version and using the `jsdom` are not affected by these bypasses. 

It appears `htmlparser2` provides different parsing results, leading to different sanitization results compared to the client-side.
I tried to looking into this but couldn't find a reliable workaround. I also tried a few other parsers (ex. parse5, xmldom, dom-parser, html5parser) but nothing really worked.

With this in mind, I'd suggest going back to using `jsdom`. Some of the key points of this decision are:
* it provides the same output and sanitization results as the original client side
* doesn't require extra custom code to get it to work with the sanitization code
* The downside I've noticed, is that it adds 10-15ms to request time, but it's probably a good tradeoff for better security and maintainability

**Comment by tyxla** [comment]
Thanks for the thorough research @claudiucelfilip! I personally agree with your sentiment and I'm happy to try `jsdom` in that case. I'd love to get a second opinion from from @Automattic/team-calypso-frameworks and @Automattic/team-calypso-platform before we go forward though.

**Comment by noahtallen** [comment]
I don't personally have any strong opinions on this one, other than whatever approach we use should be performant and secure :) 

I am curious the performance overhead -- in what scenarios do we see that 10-15ms uplift? Is it every request, or only worst case scenarios? I'm also curious, what is the point of disabling JS in the testing instructions?

**Comment by jsnajdr** [comment]
> I'm also curious, what is the point of disabling JS in the testing instructions?

With SSR the point is that you want to be sure you see the HTML rendered by the server, not the one rendered client-side in browser. If the server markup is wrong, the client is going to overwrite it with correct version very quickly. The only signal you get about something being wrong is maybe a console warning about mismatch when hydrating.

**Comment by claudiucelfilip** [comment]
> I am curious about the performance overhead -- in what scenarios do we see that 10-15ms uplift? Is it every request, or only worst-case scenarios?

@noahtallen I had another look into the profile data but it seems now there're fewer instances where sanitization is needed for the Plugin Landing pages (only for the Yoast Premium).
However, the sanitization call happens on every `/plugins` request, even for cache hits.

I pushed an [improvement](https://github.com/Automattic/wp-calypso/pull/68195/commits/c7619f6dea7ef64b3fb64f4095d4de0a82113c14) so that react-query caches the sanitized version. With this, the sanitizer kicks in only when the react-query cache is stale

File: bin/copy-production-modules.js
Comment (claudiucelfilip) [inline bin/copy-production-modules.js:104]: This was needed to allow us to skip optional dependencies, defined through [peerDependenciesMeta field](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta).

`jsdom` has an optional dependency canvas that has trouble building, resulting in the error:
```
Error: dlopen(/.../node_modules/canvas/build/Release/canvas.node, 0x0001): symbol not found in flat namespace (_cairo_fill)
```
This is most likely causing issues with TeamCity builds as well.

Fortunately, this [is an optional](https://github.com/jsdom/jsdom/blob/master/package.json#L57) dependency, and skipping it shouldn't affect the sanitizing process.

Comment (noahtallen) [reply bin/copy-production-modules.js:104]: That makes sense to me. Here's some more info about Canvas in jsdom: https://github.com/jsdom/jsdom#canvas-support. It looks like we would have to include the dependency ourselves for it to work.

File: bin/copy-production-modules.js
Comment (noahtallen) [inline bin/copy-production-modules.js]: Is this enough to detect optional peer dependencies? Should we filter for the `optional: true` property first?

File: client/data/marketplace/use-wpcom-plugins-query.ts
Comment (noahtallen) [inline client/data/marketplace/use-wpcom-plugins-query.ts:44]: Any way we could avoid `any` here? (maybe `unknown` if we really don't know the type?)

File: client/data/marketplace/use-wporg-plugin-query.ts
Comment (noahtallen) [inline client/data/marketplace/use-wporg-plugin-query.ts:27]: Same thought re: any

File: client/lib/plugins/sanitize-section-content.js
Comment (noahtallen) [inline client/lib/plugins/sanitize-section-content.js]: Just a drive-by note that we don't need to support IE11 if it simplifies things

File: client/lib/plugins/sanitize-section-content.js
Comment (noahtallen) [inline client/lib/plugins/sanitize-section-content.js:17]: One concern (but I'm not sure if it's valid) is whether this root and parser will be shared among multiple requests, and then whether that's problematic. (given it's a global value). Should we instead encapsulate it somehow?

I think I may just misunderstand what the root does :)

**Review by noahtallen — COMMENTED** [review]
Left a few questions. I'm not very familiar with this part of the code but thanks for working on it! :)

File: bin/copy-production-modules.js
Comment (noahtallen) [reply bin/copy-production-modules.js:104]: plus, if we aren't allowing the `canvas` tag in the sanitizer anyways, we wouldn't have any use for the dependency

File: client/lib/plugins/sanitize-section-content.js
Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js:17]: It's a stand-in for the `window` object. On the server, we want to replace it with the `jsdom` version.

File: client/data/marketplace/use-wpcom-plugins-query.ts
Comment (claudiucelfilip) [reply client/data/marketplace/use-wpcom-plugins-query.ts:44]: Fair point. 
I replaced it with `unknown`.

File: bin/copy-production-modules.js
Comment (claudiucelfilip) [reply bin/copy-production-modules.js]: That makes sense.
I added a filter to check for the property.

File: client/lib/plugins/sanitize-section-content.js
Comment (claudiucelfilip) [reply client/lib/plugins/sanitize-section-content.js:17]: The `root` property is [overwritten](https://github.com/Automattic/wp-calypso/pull/68195/files#diff-8c13d14907acf58f3cd93102c02ad1f74e9b4ec941636b12fab7cfdbd12e0befR14) on server start. Also, a `DOMParser` instance is created and shared between requests since we're only using its `parseFromString` method.

Client-side should continue to use `window` and create a new `DOMParser` instance each time, as before.

File: client/data/marketplace/use-wpcom-plugins-query.ts
Comment (claudiucelfilip) [reply client/data/marketplace/use-wpcom-plugins-query.ts:44]: Actually, using `unknown` start breaking type-checks in other files (such as this [one here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/marketplace/pages/marketplace-plugin-install/index.tsx#L300)).

I restored the `any` type but we'll want to follow up and define the normalized plugin type ([here](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/plugins/utils.js#L186)).

File: client/lib/plugins/sanitize-section-content.js
Comment (noahtallen) [reply client/lib/plugins/sanitize-section-content.js:17]: Makes sense, thank you for clarifying 👍

File: package.json
Comment (noahtallen) [inline package.json]: I noticed we aren't importing most of these dependencies. Are they all peer dependencies of jsdom?

**Review by noahtallen — APPROVED** [review]
Other than my comment about dependencies, this is looking pretty good to me. I think it would be a good idea to get a second approval from @jsnajdr or @tyxla on this one

File: package.json
Comment (claudiucelfilip) [reply package.json]: Indeed, these were needed because the duplicate packages were failing the TC  at the Prevent duplicated packages step. I added these based on the `yarn dedupe --check` output.
I reran it and the list has shrunk now to just `decimal.js`, `node-gyp-build` and `psl`.

File: client/lib/plugins/allowed-tags.js
Comment (tyxla) [inline client/lib/plugins/allowed-tags.js:5]: Do you have any idea how this list came to be? I see it's adapting the pre-existing one that was used to sanitize HTML but was curious if you have any insights on how it was created and whether it needs some love.

Should we get it closer to https://github.com/WordPress/wordpress-develop/blob/6.0.2/src/wp-includes/kses.php#L61? 

I definitely don't want to expand the PR further, but I wanted to get your thoughts.

File: client/lib/plugins/allowed-tags.js
Comment (tyxla) [inline client/lib/plugins/allowed-tags.js:8]: Are we intentionally removing `target`? Won't this cause any external links to work like internal ones?

File: test/client/setup-test-framework.js
Comment (tyxla) [inline test/client/setup-test-framework.js]: Any chance we can set this up only for the tests that are going to need it?

**Review by tyxla — CHANGES_REQUESTED** [review]
This appears to introduce a regression with how we fetch plugins.

When testing `/plugins` and `/plugins/woocommerce-bookings` as a logged-out user, I get this:

Screenshot 2022-10-17 at 10 52 35

Seems like this is caused by us attempting to fetch plugins for the current user (`/me/plugins`) when logged out:

Screenshot 2022-10-17 at 10 57 07

File: client/lib/plugins/allowed-tags.js
Comment (claudiucelfilip) [reply client/lib/plugins/allowed-tags.js:5]: Other than the discussion for the [initial PR](https://github.com/Automattic/wp-calypso/pull/19316), which added these rules, I'm not too familiar with the reasoning and many of the details and types of content that can pass through this.
With this in mind, I'd probably avoid making sweeping changes to this and add tags and attributes on a limited, per-need basis.

File: client/lib/plugins/allowed-tags.js
Comment (claudiucelfilip) [reply client/lib/plugins/allowed-tags.js:8]: It appears so. The intent was to keep the original [switch case](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/plugins/sanitize-section-content.js#L60) where only `href` was return true.

File: test/client/setup-test-framework.js
Comment (claudiucelfilip) [reply test/client/setup-test-framework.js]: Good point!
I found a workaround by running this part in the test directly, before JSDOM.

**Comment by claudiucelfilip** [comment]
> This appears to introduce a regression with how we fetch plugins.
> 
> When testing `/plugins` and `/plugins/woocommerce-bookings` as a logged-out user, I get this:
> 
> Seems like this is caused by us attempting to fetch plugins for the current user (`/me/plugins`) when logged out:
> 

It probably needs a rebase because it doesn't include the fix made for this in https://github.com/Automattic/wp-calypso/pull/68835.

I pushed a rebase and the problem should be fixed now.

**Review by tyxla — APPROVED** [review]
This works great in my testing! The code also looks good. 👍 

Let's keep an eye on SSR performance and monitor for errors for a few days to ensure we're not introducing any regressions.

Thank you for the hard work 🚀

File: package.json
Comment (noahtallen) [reply package.json]: Oh, interesting. Typically, you just need to run `yarn dedupe`, and it updates the lockfile. The goal is to make sure that the transitive dependencies use the same version wherever possible. Normally, you don't need to update package.json. (Typically, you only need to update package.json if we explicitly import it, or it's explicitly a peer dependency.)

**Comment by noahtallen** [comment]
Re: our peer dependency discussion, I made the following change, just to double check my theory was right:

- Removed `utf-8-validate` and `bufferutil`, because we don't appear to use them in the repo, and they aren't explicit peer dependencies of `jsdom`
- Removed the added resolutions because those can just be handled by `yarn.lock`

`yarn dedupe --check` and `yarn install` both report no errors, and as long as the tests pass, I think this should work fine. (To solve dedupe errors, you normally just have to run `yarn dedupe`, and that updates the `yarn.lock` file so that transitive dependencies use the same version)

File: yarn.lock
Comment (noahtallen) [inline yarn.lock:13146]: This is an example of a "deduped" dependency in `yarn.lock` -- anything requesting v 1.1.1 or 1.1.4 will both resolve to 1.1.4, thus avoiding having extra copies of `chown` on the file system

File: yarn.lock
Comment (noahtallen) [inline yarn.lock]: This is an example of a "deduped" dependency in yarn.lock. If anything requests any of these five versions of `debug`, yarn will resolve it to version 3.2.7, thus avoiding extra copies of the dependency on the file system. Normally, yarn handles all this automatically, but sometimes you need to force an update by running `yarn dedupe`

File: package.json
Comment (claudiucelfilip) [inline package.json]: I think we need keep "lzma-native": "^8.0.5" here, otherwise `yarn install` is starting to fail for me.
` YN0009: │ lzma-native@npm:6.0.1 couldn't be built successfully ` - it's trying to build `6.0.1` which is unable to download (404).

**Comment by claudiucelfilip** [comment]
@noahtallen Thanks for removing the unnecessary dependencies. I'm surprised to see they're not needed any longer.
I'm not sure what happened between the first time I added them and now; maybe it's from the evolving nature of the dependencies.

**Review by xknown — APPROVED** [review]
The code looks good to me. Thanks for addressing the initial feedback!

File: package.json
Comment (claudiucelfilip) [reply package.json]: The failure is happening with `6.0.1` of `lzma-native` on M1 Mac chips and has been addressed in #60959.
I pushed a commit to restore it.

Comment (noahtallen) [reply package.json]: Interesting! I have an M1 chip and it wasn't causing issues.

**Pull request merged**
